proposals/p3720.md

```diff
@@ -273,99 +285,114 @@ impl __TypeOf_C_F as RefBind(C)
 > [non-instance interface members](#non-instance-interface-members).
 
 Those implementations are how we get from `__C_F` with type `__TypeOf_C_F` to
-`v as __ValueBind_C_F` or `&r as __RefBind_C_F*`:
+`v as __Bind_C_F` or `&r as __Bind_C_F*`, conceptually following these steps:
 
 ```carbon
 // `v` is a value and so uses `ValueBind`
 v.F() == v.(C.F)()
       == v.(__C_F)()
-      == __C_F.((__TypeOf_C_F as ValueBind(C)).Op(v))()
-      == (v as __ValueBind_C_F)()
+      == __C_F.((__TypeOf_C_F as ValueBind(C)).Op)(v)()
+      == (v as __Bind_C_F)()
 
 // `r` is a reference expression and so uses `RefBind`
 r.F() == r.(C.F)()
       == r.(__C_F)()
-      == (*__C_F.((__TypeOf_C_F as RefBind(C)).Op(&r)))()
-      == (*(&r as __RefBind_C_F*))()
+      == (*__C_F.((__TypeOf_C_F as RefBind(C)).Op)(&r))()
+      == (*(&r as __Bind_C_F*))()
 ```
 
-At this point we have resolved the binding, and are left with an expression
-followed by `()`. In the first case, that expression is a value expression of
-type `__Value_Bind_C_F`. In the second case, it is a reference expression of
-type `__RefBind_C_F`.
+However, to avoid recursive application of these same rules, we need to avoid
+expressing this in terms of evaluating `__C_F.(`...`)`. Instead the third step
+shall instead use an intrinsic compiler primitive, as in:
+
+```carbon
+// `v` is a value and so uses `ValueBind`
+v.F() == v.(C.F)()
+      == v.(__C_F)()
+      == inlined_method_call_compiler_intrinsic(
+              <function body (__TypeOf_C_F as ValueBind(C)).Op overload 0>,
+              __C_F, (v))()
+      == (v as __Bind_C_F)()
+
+// `r` is a reference expression and so uses `RefBind`
+r.F() == r.(C.F)()
+      == r.(__C_F)()
+      == (*inlined_method_call_compiler_intrinsic(
```

Comment:
Should this be a primitive dereference, rather than this desugaring to [a call to `Pointer.Dereference`](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/values.md#dereferencing-customization)? I think the desugaring would not create a cycle, because we'd end up in `ValueBind` instead, but maybe we want to avoid the extra steps here if we're desugaring to primitives anyway? I would expect the implementation of `Pointer.Dereference` for pointers would be `final`.